### PR TITLE
Write evaluateScript to typecheck and apply value restriction

### PR DESCRIPTION
**Pull request by vmchale**
This modifies `evaluateScript` to typecheck things, so that `runScript` can be used by the validation server.

### CODE REVIEW

File: wallet-api/src/Ledger/Scripts.hs [hunk 1]
```diff
@@ -83,7 +85,7 @@ scriptSize (Script s) = PLC.programSize s
 
 -- | Turn a 'CompiledCode' (usually produced by 'compile') into a 'Script' for use with this package.
 fromCompiledCode :: CompiledCode a -> Script
-fromCompiledCode = Script . getPlc
+fromCompiledCode = Script . PLC.runQuote . PLC.normalizeTypesFullInProgram . getPlc
```
Comment (michaelpj) [inline wallet-api/src/Ledger/Scripts.hs]: Probably the Haddock should mention that it normalizes types?

**Comment by j-mueller** [comment]
hm, looks like there's still the `Malformed type at (). Type '(lam dat (fun (type) (type)) [dat a])' is not a neutral type` error.

The script that causes it is: `() () () -> $$(PlutusTx.traceH) "I always fail everything" (Builtins.error @())` Maybe try changing it to `() () (_ :: PendingTx) -> ...` ? The 3rd argument is a `PendingTx` value

**Comment by michaelpj** [comment]
Yeah, that looks like a type error. Weird that it's reporting a type normalization error instead...

**Comment by vmchale** [comment]
@j-mueller still fails unfortunately.

@michaelpj I think the problem is that `normalizeTypesFullIn` and `isTermValue` are not working together as they should. If `normalizeTypesFullIn` is called on a term, then it should satisfy `isTermValue`.

**Comment by michaelpj** [comment]
> I think the problem is that normalizeTypesFullIn and isTermValue are not working together as they should. If normalizeTypesFullIn is called on a term, then it should satisfy isTermValue.

That doesn't seem right - `normalizeTypes` normalizes the *types*, it doesn't normalize the term. So there's no reason to expect it to be a value afterwards.

**Comment by vmchale** [comment]
@michaelpj My mistake. It actually calls `checkProgram` under the hood.

**Comment by vmchale** [comment]
@michaelpj have a look at: https://github.com/input-output-hk/plutus/pull/925/files#diff-cf1d83c2d4eefab060bd4aaa0a12a16aR108

This is where I'm getting the normalization error so I'm puzzled.

**Comment by michaelpj** [comment]
Hard to say, I think you'd need to look at the actual script.

**Comment by vmchale** [comment]
Yes, I was hoping someone would have more insight than I :(

The script is 3700 lines of PLC so it'll take a longtime to figure out what's going on

**Comment by michaelpj** [comment]
Give this a rebase and it should be fine now, right?

**Comment by vmchale** [comment]
Not quite, unfortunately. I need to fix the typechecker to allow builtins for this particular bit.

**Comment by effectfully** [comment]
@vmchale, what's wrong with the type checker?

**Comment by vmchale** [comment]
Nothing, but as it is used here it does not include the dynamic builtins necessary to run `wallet-api` code.

**Comment by michaelpj** [comment]
Yeah, ideally we wouldn't include them, but we haven't worked out how to do the conditional inclusion of the string stuff, so I think that's the best approach at the moment.

**Comment by michaelpj** [comment]
Got this working in https://github.com/input-output-hk/plutus/pull/1119!

**Pull request closed without merging**